FreeRTOS/FreeRTOS-Kernel
# Fix portSET_INTERRUPT_MASK_FROM_ISR definition for atomic operation

Fix portSET_INTERRUPT_MASK_FROM_ISR definition for atomic operation

Description
-----------
In FreeRTOS.h, portSET_INTERRUPT_MASK_FROM_ISR is always defined to 0 if left undefined.
```c
#define portSET_INTERRUPT_MASK_FROM_ISR()       0
```

This cause the following code in atomic.h always been evaluated to true by preprocessor.
```c
#if defined( portSET_INTERRUPT_MASK_FROM_ISR )

/* Nested interrupt scheme is supported in this port. */
    #define ATOMIC_ENTER_CRITICAL() \
    UBaseType_t uxCriticalSectionType = portSET_INTERRUPT_MASK_FROM_ISR()

    #define ATOMIC_EXIT_CRITICAL() \
    portCLEAR_INTERRUPT_MASK_FROM_ISR( uxCriticalSectionType )

#else

/* Nested interrupt scheme is NOT supported in this port. */
    #define ATOMIC_ENTER_CRITICAL()    portENTER_CRITICAL()
    #define ATOMIC_EXIT_CRITICAL()     portEXIT_CRITICAL()

#endif /* portSET_INTERRUPT_MASK_FROM_ISR() */
```

In this PR
* Set `portHAS_NESTED_INTERRUPTS` to 1 if portSET_INTERRUPT_MASK_FROM_ISR and portCLEAR_INTERRUPT_MASK_FROM_ISR are defined by the port.
* Update the atomic.h to use `#if ( portHAS_NESTED_INTERRUPTS == 1 )` instead.
* Update the ports which don't support nested interrupts to remove the define of portSET/CLEAR_INTERRUPT_MASK_FROM_ISR macros. Port should not define these macros if nested interrupt is not supported.

Test Steps
-----------
Not define  `portSET/CLEAR_INTERRUPT_MASK_FROM_ISR` in portmacro.h, the following defines are declared in atomic.h:
```c
/* Nested interrupt scheme is NOT supported in this port. */
    #define ATOMIC_ENTER_CRITICAL()    portENTER_CRITICAL()
    #define ATOMIC_EXIT_CRITICAL()     portEXIT_CRITICAL()
```

Define `portSET/CLEAR_INTERRUPT_MASK_FROM_ISR` in portmacro.h, the following defines are declared in atomic.h:
```c
/* Nested interrupt scheme is supported in this port. */
    #define ATOMIC_ENTER_CRITICAL() \
    UBaseType_t uxCriticalSectionType = portSET_INTERRUPT_MASK_FROM_ISR()

    #define ATOMIC_EXIT_CRITICAL() \
    portCLEAR_INTERRUPT_MASK_FROM_ISR( uxCriticalSectionType )
```


Checklist:
----------



Related Issue
-----------
#936 


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
